{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f87942a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import sys\n",
    "sys.path.append(str(Path().resolve().parent))\n",
    "\n",
    "import pandas as pd\n",
    "from lib.genetic_data_helpers import *\n",
    "from lib.data_access_helpers import load_fasta_records, parse_gff3_tmr_file\n",
    "from lib.aminoacids import AA_1_TO_3, AA_3_TO_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a208aa8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences1 = parse_gff3_tmr_file('../data/tmhmm/TMRs_dep_1.gff3')\n",
    "sequences2 = parse_gff3_tmr_file('../data/tmhmm/TMRs_dep_2.gff3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa86cc8b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           length     sp    tmd  first_hydrophobic_start  \\\n",
      "gene                                                       \n",
      "YAL064W-B     126  False  False                      NaN   \n",
      "YAL055W       180  False  False                      NaN   \n",
      "YAL042W       415  False   True                     25.0   \n",
      "YAL023C       759  False   True                     62.0   \n",
      "YAL007C       215   True   True                      1.0   \n",
      "...           ...    ...    ...                      ...   \n",
      "YPR149W       173  False   True                      4.0   \n",
      "YPR159C-A      33  False   True                     11.0   \n",
      "YPR170W-B      85  False   True                     10.0   \n",
      "YPR198W       543  False   True                      9.0   \n",
      "YPR201W       404  False   True                     35.0   \n",
      "\n",
      "           first_hydrophobic_end  \n",
      "gene                              \n",
      "YAL064W-B                    NaN  \n",
      "YAL055W                      NaN  \n",
      "YAL042W                     45.0  \n",
      "YAL023C                     77.0  \n",
      "YAL007C                     25.0  \n",
      "...                          ...  \n",
      "YPR149W                     24.0  \n",
      "YPR159C-A                   31.0  \n",
      "YPR170W-B                   31.0  \n",
      "YPR198W                     29.0  \n",
      "YPR201W                     55.0  \n",
      "\n",
      "[435 rows x 5 columns]\n"
     ]
    }
   ],
   "source": [
    "df_dict = {\n",
    "    'gene': [],\n",
    "    'length': [],\n",
    "    'sp': [],\n",
    "    'tmd': [],\n",
    "    'first_hydrophobic_start': [],\n",
    "    'first_hydrophobic_end': []\n",
    "}\n",
    "\n",
    "sequences = sequences1 + sequences2\n",
    "\n",
    "for seq in sequences:\n",
    "    df_dict['gene'].append(seq['sequence_id'])\n",
    "    df_dict['length'].append(int(seq['length']))\n",
    "    sp = False\n",
    "    tmd = False\n",
    "    start = None\n",
    "    end = None\n",
    "    \n",
    "    for feature_dict in seq['features']:\n",
    "        if feature_dict['type'] == 'signal':\n",
    "            sp = True\n",
    "            if start is None:\n",
    "                start = int(feature_dict['start'])\n",
    "            if end is None:\n",
    "                end = int(feature_dict['end'])\n",
    "        if feature_dict['type'] == 'TMhelix':\n",
    "            tmd = True\n",
    "            if start is None:\n",
    "                start = int(feature_dict['start'])\n",
    "            if end is None:\n",
    "                end = int(feature_dict['end'])\n",
    "    \n",
    "    df_dict['sp'].append(sp)\n",
    "    df_dict['tmd'].append(tmd)                \n",
    "    df_dict['first_hydrophobic_start'].append(start)\n",
    "    df_dict['first_hydrophobic_end'].append(end)\n",
    "\n",
    "\n",
    "df = pd.DataFrame(df_dict)\n",
    "df = df.set_index('gene')\n",
    "\n",
    "df.to_csv(\"../data/tmhmm/dependent_hydrophobic_tmhmm_all.csv\", sep=\",\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fcee0624",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "344"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "########### SRP-DEPENDENT ###############\n",
    "\n",
    "# Filter the TMHMM resulting list to include only those SRP dependent genes (from Ast et al. 2013)\n",
    "# with hydrophobic regions (SP or TMD) starting in the first 60 amino acids\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "path = \"../data/tmhmm/dependent_hydrophobic_tmhmm_all.csv\"\n",
    "df = pd.read_csv(path, sep=',')\n",
    "\n",
    "sp_or_tmd = df[(df['sp'] == True) | (df['tmd'] == True)]\n",
    "\n",
    "early_hydro = sp_or_tmd[sp_or_tmd['first_hydrophobic_start'] <= 60.0]\n",
    "\n",
    "# Also, we want to remove the follwoing genes, which are too short for our window analyses.\n",
    "# These genes were identified while running the calculations\n",
    "too_short = [\n",
    "    \"YBR196C-A\",\n",
    "    \"YCL005W-A\",\n",
    "    \"YCR024C-A\",\n",
    "    \"YDL232W\",\n",
    "    \"YDL067C\",\n",
    "    \"YDR276C\",\n",
    "    \"YEL017C-A\",\n",
    "    \"YER039C-A\",\n",
    "    \"YER053C-A\",\n",
    "    \"YGL226C-A\",\n",
    "    \"YGL194C-A\",\n",
    "    \"YGR105W\",\n",
    "    \"YIL134C-A\",\n",
    "    \"YJL205C\",\n",
    "    \"YJL127C-B\",\n",
    "    \"YJR010C-A\",\n",
    "    \"YKL065W-A\",\n",
    "    \"YLR342W-A\",\n",
    "    \"YMR256C\",\n",
    "    \"YOR032W-A\",\n",
    "    \"YPL096C-A\",\n",
    "    \"YPR010C-A\",\n",
    "    \"YPR159C-A\",\n",
    "    \"YPR170W-B\"\n",
    "]\n",
    "\n",
    "filtered = early_hydro[~early_hydro['gene'].isin(too_short)]\n",
    "\n",
    "filtered.to_csv(\"../data/dependent_hydrophobic_genes.csv\", sep=',', index=False)\n",
    "\n",
    "####### Create the filtered fasta files ##########\n",
    "\n",
    "dna_path = '../data/orf_genomic_all.fasta'\n",
    "protein_path = '../data/orf_trans_all.fasta'\n",
    "\n",
    "genes = filtered[\"gene\"].tolist()\n",
    "\n",
    "# filter DNA file\n",
    "filtered_dna = []\n",
    "\n",
    "for record in SeqIO.parse(dna_path, \"fasta\"):\n",
    "    sys_name = str(record.id)\n",
    "    if sys_name in genes:\n",
    "        filtered_dna.append(record)\n",
    "        \n",
    "out_path = \"../data/dep_dna_filtered.fasta\"\n",
    "SeqIO.write(filtered_dna, out_path, \"fasta\")\n",
    "\n",
    "# filter Proteins file\n",
    "filtered_proteins = []\n",
    "\n",
    "for record in SeqIO.parse(protein_path, \"fasta\"):\n",
    "    sys_name = str(record.id)\n",
    "    if sys_name in genes:\n",
    "        filtered_proteins.append(record)\n",
    "        \n",
    "out_path = \"../data/dep_proteins_filtered.fasta\"\n",
    "SeqIO.write(filtered_proteins, out_path, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "edb20803",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "299"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "########## SRP-INDEPENDENT ##############\n",
    "\n",
    "# Filter the TMHMM resulting list to include only those SRP INdependent genes (from Ast et al. 2013)\n",
    "# with hydrophobic regions (SP or TMD) starting in the first 60 amino acids\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "path = \"../data/tmhmm/independent_hydrophobic_tmhmm_all.csv\"\n",
    "df = pd.read_csv(path, sep=',')\n",
    "\n",
    "sp_or_tmd = df[(df['sp'] == True) | (df['tmd'] == True)]\n",
    "\n",
    "early_hydro = sp_or_tmd[sp_or_tmd['first_hydrophobic_start'] <= 60.0]\n",
    "\n",
    "too_short = [\n",
    "    \"YAR020C\",\n",
    "    \"YDR119W-A\",\n",
    "    \"YDR182W-A\",\n",
    "    \"YDR524C-B\",\n",
    "    \"YJL052C-A\",\n",
    "    \"YLR406C-A\",\n",
    "    \"YMR251W-A\",\n",
    "    \"YNL024C-A\",\n",
    "    \"YOL052C-A\"\n",
    "]\n",
    "\n",
    "filtered = early_hydro[~early_hydro['gene'].isin(too_short)]\n",
    "\n",
    "filtered.to_csv(\"../data/independent_hydrophobic_genes.csv\", sep=',', index=False)\n",
    "\n",
    "####### Create the filtered fasta files ##########\n",
    "\n",
    "dna_path = '../data/orf_genomic_all.fasta'\n",
    "protein_path = '../data/orf_trans_all.fasta'\n",
    "\n",
    "genes = filtered[\"gene\"].tolist()\n",
    "\n",
    "# filter DNA file\n",
    "filtered_dna = []\n",
    "\n",
    "for record in SeqIO.parse(dna_path, \"fasta\"):\n",
    "    sys_name = str(record.id)\n",
    "    if sys_name in genes:\n",
    "        filtered_dna.append(record)\n",
    "        \n",
    "out_path = \"../data/indep_dna_filtered.fasta\"\n",
    "SeqIO.write(filtered_dna, out_path, \"fasta\")\n",
    "\n",
    "# filter Proteins file\n",
    "filtered_proteins = []\n",
    "\n",
    "for record in SeqIO.parse(protein_path, \"fasta\"):\n",
    "    sys_name = str(record.id)\n",
    "    if sys_name in genes:\n",
    "        filtered_proteins.append(record)\n",
    "        \n",
    "out_path = \"../data/indep_proteins_filtered.fasta\"\n",
    "SeqIO.write(filtered_proteins, out_path, \"fasta\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
