{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "04e129b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import sys\n",
    "sys.path.append(str(Path().resolve().parent))\n",
    "\n",
    "from Bio import SeqIO\n",
    "from lib.calculations import run_rscu_analysis\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from lib.aminoacids import SLOW, FAST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cb91b7dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "######## differences metrics #############\n",
    "\n",
    "def sum_of_abs_differences_fast_slow(rscu_w, rscu_b, codons):\n",
    "    diffs = abs(rscu_w - rscu_b)\n",
    "    slow_sum = sum(diffs[i] for i, codon in enumerate(codons) if codon in SLOW)\n",
    "    fast_sum = sum(diffs[i] for i, codon in enumerate(codons) if codon in FAST)\n",
    "    return slow_sum, fast_sum\n",
    "\n",
    "def sum_dist_from_line(rscu_w, rscu_b, codons):\n",
    "    dists = rscu_b - rscu_w\n",
    "    slow_sum = sum(dists[i] for i, codon in enumerate(codons) if codon in SLOW)\n",
    "    fast_sum = sum(dists[i] for i, codon in enumerate(codons) if codon in FAST)\n",
    "    return slow_sum, fast_sum\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "11b48ad5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "------------------------------\n",
      "Best parameters found for SLOW CODONS:\n",
      "Best window size (in amino acids): 45, Best distance from CS (in amino acids): 90, Max sum of differences: 0.113\n",
      "------------------------------\n",
      "Best parameters found for FAST CODONS:\n",
      "Best window size (in amino acids): 20, Best distance from CS (in amino acids): 80, Min sum of differences: -0.323\n"
     ]
    }
   ],
   "source": [
    "###############  Maximize Slow codon RSCU and minimize Fast codon RSCU  ###################\n",
    "\n",
    "UseCs = True\n",
    "UseSpAsWindow = False\n",
    "        \n",
    "# load SP ORFS (genomic)\n",
    "all_fasta_path = '../data/dep_dna_filtered.fasta'\n",
    "records = list(SeqIO.parse(all_fasta_path, 'fasta'))\n",
    "\n",
    "# load SP data\n",
    "tsv_path = '../data/dependent_hydrophobic_genes.csv'\n",
    "df = pd.read_csv(tsv_path, index_col='gene')\n",
    "\n",
    "# build numpy array of cleavage site values corresponding to the ordering of records\n",
    "arr_cs_aa = np.zeros(len(records))\n",
    "for i, record in enumerate(records):\n",
    "    row = df.loc[record.id]\n",
    "    arr_cs_aa[i] = row['first_hydrophobic_end'] if UseCs else 0\n",
    "arr_cs_n = arr_cs_aa * 3\n",
    "\n",
    "max_diff_slow = 0\n",
    "best_w_aa_slow = 0\n",
    "best_d_aa_slow = 0\n",
    "\n",
    "min_diff_fast = 0\n",
    "best_w_aa_fast = 0\n",
    "best_d_aa_fast = 0\n",
    "\n",
    "for w_aa in range(20, 50, 5): # searching from window size (in amino acids) from 10, 15, 20, ...,50\n",
    "    for d_aa in range(0, 100, 10): # searching in distance from CS (in amino acids) from 0, 10, 20, ..., 100\n",
    "        \n",
    "        w_n = w_aa * 3\n",
    "        arr_npet_n = np.full(len(records), d_aa * 3, dtype=int)\n",
    "        \n",
    "        window_rscu, comp_rscu = run_rscu_analysis(records, arr_cs_n, arr_npet_n, w_n, UseCs, UseSpAsWindow, silent=True)\n",
    "        \n",
    "        rscu_w = np.array(list(window_rscu.values()))\n",
    "        rscu_b = np.array(list(comp_rscu.values()))\n",
    "        \n",
    "        codons = list(window_rscu.keys())\n",
    "        \n",
    "        slow_sum, fast_sum = sum_dist_from_line(rscu_w, rscu_b, codons)\n",
    "        \n",
    "        if slow_sum > max_diff_slow:\n",
    "            max_diff_slow = slow_sum\n",
    "            best_w_aa_slow = w_aa\n",
    "            best_d_aa_slow = d_aa\n",
    "            \n",
    "        if fast_sum < min_diff_fast:\n",
    "            min_diff_fast = fast_sum\n",
    "            best_w_aa_fast = w_aa\n",
    "            best_d_aa_fast = d_aa\n",
    "            \n",
    "            \n",
    "   \n",
    "print(\"\")\n",
    "print(\"------------------------------\")\n",
    "print(\"Best parameters found for SLOW CODONS:\")         \n",
    "print(f\"Best window size (in amino acids): {best_w_aa_slow}, Best distance from CS (in amino acids): {best_d_aa_slow}, Max sum of differences: {max_diff_slow:.3f}\")\n",
    "\n",
    "print(\"------------------------------\")\n",
    "print(\"Best parameters found for FAST CODONS:\")\n",
    "print(f\"Best window size (in amino acids): {best_w_aa_fast}, Best distance from CS (in amino acids): {best_d_aa_fast}, Min sum of differences: {min_diff_fast:.3f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "56a8922d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "------------------------------\n",
      "Best parameters found for SLOW CODONS:\n",
      "Best window size (in amino acids): 20, Best distance from CS (in amino acids): 10, Max sum of abs differences: 1.327\n",
      "------------------------------\n",
      "Best parameters found for FAST CODONS:\n",
      "Best window size (in amino acids): 25, Best distance from CS (in amino acids): 40, Max sum of abs differences: 1.104\n",
      "------------------------------\n",
      "Best parameters found overall:\n",
      "Best window size (in amino acids): 20, Best distance from CS (in amino acids): 0, Max sum of abs differences: 2.330\n"
     ]
    }
   ],
   "source": [
    "###############  Maximize absoslute difference  ###################\n",
    "\n",
    "UseCs = True\n",
    "UseSpAsWindow = False\n",
    "        \n",
    "# load SP ORFS (genomic)\n",
    "all_fasta_path = '../data/dep_dna_filtered.fasta'\n",
    "records = list(SeqIO.parse(all_fasta_path, 'fasta'))\n",
    "\n",
    "# load SP data\n",
    "tsv_path = '../data/dependent_hydrophobic_genes.csv'\n",
    "df = pd.read_csv(tsv_path, index_col='gene')\n",
    "\n",
    "# build numpy array of cleavage site values corresponding to the ordering of records\n",
    "arr_cs_aa = np.zeros(len(records))\n",
    "for i, record in enumerate(records):\n",
    "    row = df.loc[record.id]\n",
    "    arr_cs_aa[i] = row['first_hydrophobic_end'] if UseCs else 0\n",
    "arr_cs_n = arr_cs_aa * 3\n",
    "\n",
    "max_diff_slow = 0\n",
    "best_w_aa_slow = 0\n",
    "best_d_aa_slow = 0\n",
    "\n",
    "max_diff_fast = 0\n",
    "best_w_aa_fast = 0\n",
    "best_d_aa_fast = 0\n",
    "\n",
    "max_diff_overall = 0\n",
    "best_w_aa_overall = 0\n",
    "best_d_aa_overall = 0\n",
    "\n",
    "for w_aa in range(20, 50, 5): # searching from window size (in amino acids) from 10, 15, 20, ...,50\n",
    "    for d_aa in range(0, 100, 10): # searching in distance from CS (in amino acids) from 0, 10, 20, ..., 100\n",
    "        \n",
    "        w_n = w_aa * 3\n",
    "        arr_npet_n = np.full(len(records), d_aa * 3, dtype=int)\n",
    "        \n",
    "        window_rscu, comp_rscu = run_rscu_analysis(records, arr_cs_n, arr_npet_n, w_n, UseCs, UseSpAsWindow, silent=True)\n",
    "        \n",
    "        rscu_w = np.array(list(window_rscu.values()))\n",
    "        rscu_b = np.array(list(comp_rscu.values()))\n",
    "        \n",
    "        codons = list(window_rscu.keys())\n",
    "        \n",
    "        slow_sum, fast_sum = sum_of_abs_differences_fast_slow(rscu_w, rscu_b, codons)\n",
    "        \n",
    "        if slow_sum > max_diff_slow:\n",
    "            max_diff_slow = slow_sum\n",
    "            best_w_aa_slow = w_aa\n",
    "            best_d_aa_slow = d_aa\n",
    "            \n",
    "        if fast_sum > max_diff_fast:\n",
    "            max_diff_fast = fast_sum\n",
    "            best_w_aa_fast = w_aa\n",
    "            best_d_aa_fast = d_aa\n",
    "            \n",
    "        if slow_sum + fast_sum > max_diff_overall:\n",
    "            max_diff_overall = slow_sum + fast_sum\n",
    "            best_w_aa_overall = w_aa\n",
    "            best_d_aa_overall = d_aa\n",
    "            \n",
    "            \n",
    "   \n",
    "print(\"\")\n",
    "print(\"------------------------------\")\n",
    "print(\"Best parameters found for SLOW CODONS:\")         \n",
    "print(f\"Best window size (in amino acids): {best_w_aa_slow}, Best distance from CS (in amino acids): {best_d_aa_slow}, Max sum of abs differences: {max_diff_slow:.3f}\")\n",
    "\n",
    "print(\"------------------------------\")\n",
    "print(\"Best parameters found for FAST CODONS:\")\n",
    "print(f\"Best window size (in amino acids): {best_w_aa_fast}, Best distance from CS (in amino acids): {best_d_aa_fast}, Max sum of abs differences: {max_diff_fast:.3f}\")\n",
    "\n",
    "print(\"------------------------------\")\n",
    "print(\"Best parameters found overall:\")\n",
    "print(f\"Best window size (in amino acids): {best_w_aa_overall}, Best distance from CS (in amino acids): {best_d_aa_overall}, Max sum of abs differences: {max_diff_overall:.3f}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
